core/src/main/scala/kafka/coordinator/transaction/TransactionMetadata.scala

```diff
@@ -295,8 +307,16 @@ private[transaction] class TransactionMetadata(val transactionalId: String,
             txnStartTimestamp = transitMetadata.txnStartTimestamp
             topicPartitions.clear()
           }
+        case Dead => // from expire transactionalId
+          if(producerEpoch != transitMetadata.producerEpoch)
+            throw new IllegalStateException(s"Completing transaction state transition to state: $toState for transactionalId: " +
+              s"$transactionalId, producerId: $producerId, failed due to producerEpoch: $producerEpoch != expected producerEpoch: ${transitMetadata.producerEpoch}")
       }
 
+      if (toState != Dead)
+        txnLastUpdateTimestamp = transitMetadata.txnLastUpdateTimestamp
```

Comment:
Yeah i should have! Will remove anyway as we aren't going to transition to `Dead`